client/internal/routemanager/dynamic/route.go

```diff
@@ -0,0 +1,359 @@
+package dynamic
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/netip"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	log "github.com/sirupsen/logrus"
+
+	nberrors "github.com/netbirdio/netbird/client/errors"
+	"github.com/netbirdio/netbird/client/internal/peer"
+	"github.com/netbirdio/netbird/client/internal/routemanager/refcounter"
+	"github.com/netbirdio/netbird/client/internal/routemanager/util"
+	"github.com/netbirdio/netbird/management/domain"
+	"github.com/netbirdio/netbird/route"
+)
+
+const (
+	DefaultInterval = time.Minute
+
+	minInterval = 2 * time.Second
+
+	addAllowedIP = "add allowed IP %s: %w"
+)
+
+type domainMap map[domain.Domain][]netip.Prefix
+
+type resolveResult struct {
+	domain domain.Domain
+	prefix netip.Prefix
+	err    error
+}
+
+type Route struct {
+	route                *route.Route
+	routeRefCounter      *refcounter.RouteRefCounter
+	allowedIPsRefcounter *refcounter.AllowedIPsRefCounter
+	interval             time.Duration
+	dynamicDomains       domainMap
+	mu                   sync.Mutex
+	currentPeerKey       string
+	cancel               context.CancelFunc
+	statusRecorder       *peer.Status
+}
+
+func NewRoute(
+	rt *route.Route,
+	routeRefCounter *refcounter.RouteRefCounter,
+	allowedIPsRefCounter *refcounter.AllowedIPsRefCounter,
+	interval time.Duration,
+	statusRecorder *peer.Status,
+) *Route {
+	return &Route{
+		route:                rt,
+		routeRefCounter:      routeRefCounter,
+		allowedIPsRefcounter: allowedIPsRefCounter,
+		interval:             interval,
+		dynamicDomains:       domainMap{},
+		statusRecorder:       statusRecorder,
+	}
+}
+
+func (r *Route) String() string {
+	s, err := r.route.Domains.String()
+	if err != nil {
+		return r.route.Domains.PunycodeString()
+	}
+	return s
+}
+
+func (r *Route) AddRoute(ctx context.Context) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.cancel != nil {
+		r.cancel()
+	}
+
+	ctx, r.cancel = context.WithCancel(ctx)
+
+	go r.startResolver(ctx)
+
+	return nil
+}
+
+// RemoveRoute will stop the dynamic resolver and remove all dynamic routes.
+// It doesn't touch allowed IPs, these should be removed separately and before calling this method.
+func (r *Route) RemoveRoute() error {
```

Comment:
I think an inconsistent state could happen after the RemoveRoute if we call it at the wrong moment.
The **RemoveRoute** and the **r.updateDynamicRoutes** can run in two different go routine. If the updateDynamicRoutes already started to run but it did not lock yet the mutex then it will wait until the end of the RemoveRoute and continue running.

Reply:
Yep, this is missing a context done check. Fixed in 81b93e2064946a49e846109f1394b346da356483